torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -614,6 +600,7 @@ def free_unsharded_param(self) -> None:
         ):
             free_storage(tensor)
         if ca.compiled_autograd_enabled:
+            free_storage(self._unsharded_param)
```

Comment:
basic question: I thought that  `self._unsharded_param` is a view off of `self.all_gather_outputs` (the padded buffer that we allgather into)? If that's true, then isn't the `free_storage(tensor)` line above already doing the storage freeing?

Reply:
I think for arbitrary all-gather transforms the user could do an out-of-place op in the `fsdp_post_all_gather` to make the all-gather output and unsharded parameter no longer alias (e.g. dequantization).

Reply:
Should we walk through the exact pros/cons of the sacrificial param some more, just to agree on the tradeoffs of that approach vs this one? In my head, it's something like:

In the normal FSDP2 flow, for every param, we have 3 buffers: `sharded_param`, `unsharded_param`, `all_gather_output` buffer (where `unshareded_param = all_gather_output.as_strided(...)`), and `unsharded_param` is the actual "nn.Parameter" that we compute gradients on.

That means that normally, the graph signature would look something like this (I write this with 50% confidence, let me know if it looks wrong haha):
```
def fsdp_layer(sharded_param, unsharded_param, all_gather_output):
    # resizing + copy-in, to fill the all_gather_output (and its aliased unsharded_param) with data
    resize_storage(all_gather_output, full_size)
    all_gather(sharded_param, out=all_gather_output)
    # unsharded_param now has the right size + data, do compute on it
    fw_out = normal_compute(unsharded_param)
    # free the ag_buffer storage
    resize_storage(all_gather_output, 0)
    return fw_out
```

Where:

* we have lots of graph inputs, mutation and aliasing going on
* in theory, the only tensor that **needs** to be a graph input is the sharded param (the ag_output can be generated in the graph when we do an allgather
* the main problem, though, is that the **unsharded_param** is what we compute gradients on, and so it needs to be a graph input

And in theory, we could "fix" this problem by making the eager FSDP2 actually accumulate gradients directly on the sharded param. But pending that / if this is difficult to get working properly in eager, "sacrificial param" lets us accomplish the same thing by using a dummy tensor + `aten.set_`.

So forward graph when using the sacrificial param (if memory serves me) is something like:
```
def fsdp_graph_with_sacrificial_param(sharded_param, dummy):
    ag_buffer = all_gather(sharded_param)
    unsharded_param = ag_buffer.as_strided(...)
    # only purpose of the set_() is so that in the backward, we can propagate gradients to the dummy
    # now, set_() is the only mutation that we have in the forward.
    # less mutation == more opportunities for inductor to safely fuse
    dummy.set_(unsharded_param)
    #  use the unsharded param in normal fw compute
    fw_out = normal_compute(unsharded_param)
    # I think we still need a storage_resize_ op here?
    # autograd will still save the (intermediate buffer) unsharded_param for backward,
    # so we need to make sure it doesn't actually save the full tensor data
    storage_resize_(unsharded_param, 0)
    return fw_out
```

Pro:
* we have a `set_()` in the graph to deal with, but we no longer have any aliases of graph inputs in the forward.

Cons:
* We still... need to resize_storage_(0) the unsharded param in the forward, even though it is no longer a graph input, I think? (to prevent autograd from actually saving the massive buffer)

* Thinking more about our offline discussion from last Friday, I wonder if the issue has more to do with the backward:

```
def backward_graph_with_sacrifical_param(weight_activation, weight_from_bw_hook, sharded_param_from_hook, dummy_grad_out, tangent):
    allgather_buffer = allgather(sharded_param_from_hook)
    fsdp.set_(weight_from_bw_hook, allgather_buffer)
    # run the backward, which uses "weight_activation
    grad_inp = torch.matmul(tangent, weight_activation)
```

Where the problem (if I understand properly) is that:
(1) sacrificial param logic doesn't save us from the "lots of unsharded tensor inputs to the backward graph that are aliased"
(2) we were just using `fsdp.set_()` directly to update the unsharded param. But since our activation is an alias of that param, we weren't actually updating the alias

I can't think of an obvious/easy solution to those problems. So if we know we're stuck with a backward graph that has aliasing/mutation anyway, then maybe we are just stuck with having to handle that properly in inductor (and we need the corresponding scary-looking graph passes), and so we may as well add similar logic to the forward instead of using the sacrificial param.

Just trying to brainstorm a bit more on the "can we make the backward graph situation any better": I wonder if we can arrange for something like:

(1) the backward hook directly closes over the unsharded_param (and not the all_gather_output), resizing it up and doing an allgather + copy. I think... the all_gather_output shouldn't actually be necessary at this point, since the `unsharded_param` is a strided view with the right size metadata? Maybe the weird bit is that the `allgather`'s output and the `unsharded_param` that we want to copy into have different sizes, but we secretly know it's ok to do the copy since the storage sizes are the same.

(2) we rely on Horace's partitioner changes to never save views for backward, to guarantee that the "saved activation" is exactly the same tensor as the unsharded_param that the hook closes over

(3) this allows dynamo's existing "input deduping" logic to run, so dynamo will automatically dedup the param from the hook and the param saved by autograd

And then the final graph you get in the backward looks (roughly, something) like:

```
# main advantange: there are zero graph inputs that alias each other
def (hook_sharded_param, hook_unsharded_param, fw_grad_out):
    # do the same allgather + resize + split_with_sizes_copy.out that eager does
    storage_resize_(hook_unsharded_param, ag_buffer_size)
    ag_out = allgather(hook, sharded_param)
    split_with_sizes_copy.out(ag_out, out=[hook_unsharded_param])
    # this becomes the gradient for the "dummy" tensor, through the sacrificial param logic
    grad_inp = normal_backward(fw_grad_out, hook_unsharded_param)
    # inductor still needs to manually resize the storage of hook_unsharded_param,
    # since our graph previously resized it up
    storage_resize_(hook_unsharded_param, 0)
    return None, grad_inp
```

Where now, at least, we don't have any aliasing in the backward graph inputs, which might make inductor's passes less fragile. And then we can independently reason about whether/how to functionalize the above graph (which in theory should be easier given the lack of aliasing).
which we can then functionalize into this graph:
```
def (hook_sharded_param, hook_unsharded_param, fw_grad_out):
    # do the same allgather + resize + split_with_sizes_copy.out that eager does
    storage_resize_(hook_unsharded_param, ag_buffer_size)
    ag_out = allgather(hook, sharded_param)
    split_with_sizes_copy.out(ag_out, out=[hook_unsharded_param])
    # this becomes the gradient for the "dummy" tensor, through the sacrificial param logic
    grad_inp = normal_backward(fw_grad_out, hook_unsharded_param)
    # inductor still needs to manually resize the storage of hook_unsharded_param,
    # since our graph previously resized it up
    storage_resize_(hook_unsharded_param, 0)
    return None, grad_inp
```

Anyway, this is all just me trying to make sense of the current state / potentially brainstorming alternatives (which again might be wrong). If you think the current approach is the best way forward (keep the same set of inputs to the graph as in eager), I can try to review the PR more.

Reply:
oh I guess the example a user gave here is exactly that lol (doing a transform that involves compute in those hooks) https://github.com/pytorch/pytorch/issues/114299#issuecomment-2298692762

Reply:
@bdhirsh re. forward graph
> We still... need to resize_storage_(0) the unsharded param in the forward, even though it is no longer a graph input, I think? (to prevent autograd from actually saving the massive buffer)

We already guarantee that we don't have that massive temporary buffer as graph output (we explicitly free the `self.all_gather_outputs` list at end of forward, so that this list of graph-intermediate temporary buffers will not be in graph output): https://github.com/pytorch/pytorch/blob/e8fc1e0118da85c3529f71fa60823fdbbbebee56/torch/distributed/_composable/fsdp/_fsdp_param.py#L616-L618

re. backward graph
> I wonder if we can arrange for something like:
(1) the backward hook directly closes over the unsharded_param (and not the all_gather_output), resizing it up and doing an allgather + copy. I think... the all_gather_output shouldn't actually be necessary at this point, since the unsharded_param is a strided view with the right size metadata? Maybe the weird bit is that the allgather's output and the unsharded_param that we want to copy into have different sizes, but we secretly know it's ok to do the copy since the storage sizes are the same.
(2) we rely on Horace's partitioner changes to never save views for backward, to guarantee that the "saved activation" is exactly the same tensor as the unsharded_param that the hook closes over
(3) this allows dynamo's existing "input deduping" logic to run, so dynamo will automatically dedup the param from the hook and the param saved by autograd

Yes this is the intent and is how we want it to work. Unfortunately right now even if we do all these, we still see aliases in the compiled autograd graph. To provide a glimpse of the symptom (this is after switching from `.set_` to `.copy_`):
```python
Found aliased graph inputs: [arg1_1, arg28_1],
type(val): [<class 'torch._subclasses.fake_tensor.FakeTensor'>, <class 'torch._subclasses.fake_tensor.FakeTensor'>],
val.shape: [torch.Size([4096]), torch.Size([4096])],
id(val): [140576283153648, 140576283154048],
sid: 140576432772816


Found aliased graph inputs: [arg2_1, arg29_1],
type(val): [<class 'torch._subclasses.fake_tensor.FakeTensor'>, <class 'torch._subclasses.fake_tensor.FakeTensor'>],
val.shape: [torch.Size([32000, 4096]), torch.Size([32000, 4096])],
id(val): [140576283153808, 140576283153728],
sid: 140576432772256


Found aliased graph inputs: [arg3_1, arg24_1],
type(val): [<class 'torch._subclasses.fake_tensor.FakeTensor'>, <class 'torch._subclasses.fake_tensor.FakeTensor'>],
val.shape: [torch.Size([4096, 4096]), torch.Size([4096, 4096])],
id(val): [140576283152208, 140576283152048],
sid: 140576432772656
```
(tlparse: https://interncache-all.fbcdn.net/manifold/tlparse_reports/tree/logs/.tmpRzAyXI/rank_0/index.html)
We can see that each unsharded parameter has a alias of itself (exact same shape and same storage).

I tried to figure out why this happens - I found that both aliases are created in this `snapshot_fake` path: https://gist.github.com/yf225/a9c81909fc13789150d4f46a31b13039 via `ret = x.detach()` (conceptually). I believe we need a way to deduplicate those two FakeTensors when we observe they have the same shape and storage.

I put the full repro here: https://github.com/pytorch/pytorch/pull/134063